contracts/transmuter/src/test/cases/comprehensive_flows.rs

```diff
@@ -0,0 +1,797 @@
+                }],
+                token_out_min_amount: Uint128::from(1500u128).to_string(),
+            },
+            &t.accounts["alice"],
+        )
+        .unwrap_err();
+
+    assert_contract_err(
+        ContractError::InsufficientPoolAsset {
+            required: Coin::new(1_500, COSMOS_USDC),
+            available: Coin::new(0, COSMOS_USDC),
+        },
+        err,
+    );
+
+    // join pool properly
+    t.contract
+        .execute(&ExecMsg::JoinPool {}, &tokens_in, &t.accounts["provider"])
+        .unwrap();
+
+    // transmute with incorrect funds should still fail
+    let err = cp
+        .swap_exact_amount_in(
+            MsgSwapExactAmountIn {
+                sender: t.accounts["alice"].address(),
+                token_in: Some(Coin::new(1_000, COSMOS_USDC).into()),
+                routes: vec![SwapAmountInRoute {
+                    pool_id: t.contract.pool_id,
+                    token_out_denom: "urandom".to_string(),
+                }],
+                token_out_min_amount: Uint128::from(1_000u128).to_string(),
+            },
+            &t.accounts["alice"],
+        )
+        .unwrap_err();
+
+    assert_contract_err(
+        ContractError::InvalidTransmuteDenom {
+            denom: "urandom".to_string(),
+            expected_denom: vec![ETH_USDC.to_string(), COSMOS_USDC.to_string()],
+        },
+        err,
+    );
+
+    let err = cp
+        .swap_exact_amount_in(
+            MsgSwapExactAmountIn {
+                sender: t.accounts["alice"].address(),
+                token_in: Some(Coin::new(1_000, COSMOS_USDC).into()),
+                routes: vec![SwapAmountInRoute {
+                    pool_id: t.contract.pool_id,
+                    token_out_denom: "urandom2".to_string(),
+                }],
+                token_out_min_amount: Uint128::from(1_000u128).to_string(),
+            },
+            &t.accounts["alice"],
+        )
+        .unwrap_err();
+
+    assert_contract_err(
+        ContractError::InvalidTransmuteDenom {
+            denom: "urandom2".to_string(),
+            expected_denom: vec![ETH_USDC.to_string(), COSMOS_USDC.to_string()],
+        },
+        err,
+    );
+
+    // swap with correct token_in should succeed this time
+    let token_in = Coin::new(1_500, ETH_USDC);
+    cp.swap_exact_amount_in(
+        MsgSwapExactAmountIn {
+            sender: t.accounts["alice"].address(),
+            token_in: Some(token_in.into()),
+            routes: vec![SwapAmountInRoute {
+                pool_id: t.contract.pool_id,
+                token_out_denom: COSMOS_USDC.to_string(),
+            }],
+            token_out_min_amount: Uint128::from(1_500u128).to_string(),
+        },
+        &t.accounts["alice"],
+    )
+    .unwrap();
+
+    // check balances
+    t.assert_contract_balances(&[
+        Coin::new(1_500, ETH_USDC),
+        Coin::new(100_000 + 100_000 - 1_500, COSMOS_USDC), // +100_000 due to bank send
```

Comment:
I found this initially confusing. Not release blocking, but long term it would be nice to have bank hooks that allow us to execute the contract when receiving a bank send and have those tokens be added to the pool

Reply:
Isn't that what CosmWasm tries to remove due to security concerns (`recieve` in solidity)?